lib/proxy/signal_handler.rb

```diff
@@ -14,4 +23,22 @@ def self.install_ttin_trap
       end
     end
   end
+
+  def install_int_trap
+    if Rack.release < '1.6.4'
+      # Rack installs its own trap; Sleeping for 5 secs insures we overwrite it with our own
+      Thread.new do
+        sleep 5
+        begin
+          trap(:INT) { exit(0) }
+        rescue Exception => e
+          logger.warn "Unable to overwrite interrupt trap: #{e}"
+        end
+      end
+    end
+  end
+
+  def install_term_trap
+    trap(:TERM) { exit(0) }
+  end
```

Comment:
Looks like webrick is `.shutdown`-able but Rack does not seem to expose this API.

Another trick is to create some own API that will cause clean shutdown. We can generate and remember a security token and this will be passed along with the request.